packages/alfa-table/src/table.ts

```diff
@@ -1,684 +1,854 @@
+      function advance(): Option<Element> {
+        children = children.rest();
+        return current();
+      }
     }
 
-    public findGroupHeaders(): Builder {
-      const rowGroupHeaders: Array<Cell.Builder> = [];
-      const columnGroupHeaders: Array<Cell.Builder> = [];
+    /**
+     * For all downwards growing cells, extend their height to the current
+     * y-coordinate.
+     *
+     * @see https://html.spec.whatwg.org/#algorithm-for-growing-downward-growing-cells
+     */
+    function growCells(): void {
+      for (let [cell, i] of downwardGrowing) {
+        const height = yCurrent - cell.y + 1;
 
-      for (const cell of this._cells) {
-        if (cell.variant.equals(Some.of(Scope.RowGroup))) {
-          rowGroupHeaders.push(cell);
+        for (let x = cell.x, n = x + cell.width; x < n; x++) {
+          for (let y = cell.y + cell.height, n = cell.y + height; y < n; y++) {
+            add(x, y, i);
+          }
         }
-        if (cell.variant.equals(Some.of(Scope.ColumnGroup))) {
-          columnGroupHeaders.push(cell);
+
+        if (cell.isHeader()) {
+          cells[i] = Cell.header(
+            cell.element,
+            cell.anchor,
+            cell.width,
+            cell.height
+          );
+        } else {
+          cells[i] = Cell.data(cell.element, cell.anchor, cell.width, height);
         }
       }
-
-      return this.update({ rowGroupHeaders, columnGroupHeaders });
     }
 
     /**
-     * If principal cell is in a group, get all group headers that are in this group and above+lift of principal cell.
+     * @see https://html.spec.whatwg.org/#algorithm-for-assigning-header-cells
      */
-    public getAboveLeftGroupHeaders(
-      kind: "row" | "column"
-    ): (principalCell: Cell.Builder) => Iterable<Cell.Builder> {
-      let anchor: "x" | "y",
-        groups: Iterable<RowGroup | ColumnGroup>,
-        groupHeaders: Iterable<Cell.Builder>;
-
-      switch (kind) {
-        case "row":
-          anchor = "y";
-          groups = this.rowGroups;
-          groupHeaders = this._rowGroupHeaders;
-          break;
-        case "column":
-          anchor = "x";
-          groups = this.colGroups;
-          groupHeaders = this._columnGroupHeaders;
-          break;
-      }
+    function assignHeaders(cell: Cell): Cell {
+      // 1
+      const headers: Array<Cell> = [];
 
-      return (principalCell) => {
-        // The group covering the same anchor as the principal cell
-        const principalGroup = Iterable.find(groups, (group) =>
-          group.isCovering(principalCell.anchor[anchor])
+      // 2
+      // Nothing to do
+
+      // 3
+      const ids = cell.element
+        .attribute("headers")
+        .map((attribute) => attribute.tokens());
+
+      if (ids.isSome()) {
+        // 3.1
+        // Nothing to do
+
+        // 3.2
+        headers.push(
+          ...ids
+            .get()
+            .collect((id) => Option.from(cells.find(hasElement(hasId(id)))))
```

Comment:
If the table is huge, this has potential for taking a lot of time (especially if header is on the last row). Could use some cache (not sure if many cells will have the same explicit header or not).
Or maybe a pre-processing to build a map id => cell once and do quick lookup afterwards.

Reply:
That's a great suggestion 👌